source/texture3.ts

```diff
@@ -0,0 +1,400 @@
+                    document.createElementNS('http://www.w3.org/1999/xhtml', 'aux-canvas') as HTMLCanvasElement;
+                const auxiliaryContext = auxiliaryCanvas.getContext('2d');
+
+                assert(auxiliaryContext !== undefined, `expected auxiliary 2D context for temporary slicing`);
+                assert(image.height % slices === 0 && image.height / slices >= 1.0,
+                    `expected height to be a multitude of number of slices`);
+
+                const width = image.width;
+                const height = Math.floor(image.height / slices);
+                const depth = height * slices;
+
+                auxiliaryCanvas.width = width;
+                auxiliaryCanvas.height = height;
+
+                const data = new Uint8Array(width * height * depth);
+                const subImageSize = width * height;
+
+                for (let slice = 0; slice < slices; ++slice) {
+                    auxiliaryContext!.drawImage(image, 0, height * slice, width, height, 0, 0, width, height);
+                    const subImageData = auxiliaryContext!.getImageData(0, 0, width, height).data;
+
+                    for (let i = 0; i < subImageSize; ++i) {
+                        data[subImageSize * slice + i] = subImageData[4 * i];
+                    }
+                }
+
+                this.resize(width, height, depth);
+                this.data(image);
+
+                resolve();
+            };
+
+            if (crossOrigin) {
+                image.crossOrigin = 'anonymous';
+            }
+            image.src = uri;
+        });
+    }
+
+    /**
+     * Pass image data to the texture object.
+     * @param data - Texel data that will be copied into the objects data store.
+     * @param bind - Allows to skip binding the texture (e.g., when binding is handled outside).
+     * @param unbind - Allows to skip unbinding the texture (e.g., when binding is handled outside).
+     */
+    @Initializable.assert_initialized()
+    data(data: TexImage3DData, bind: boolean = true, unbind: boolean = true): void {
+        const gl = this.context.gl;
+        const gl2facade = this._context.gl2facade;
+
+        if (bind) {
+            this.bind();
+        }
+
+        gl2facade.texImage3D(gl.TEXTURE_3D, 0, this._internalFormat,
+            this._width, this._height, this._depth, 0, this._format, this._type, data);
+
+        if (unbind) {
+            this.unbind();
+        }
+        this.reallocate();
+    }
+
+    /**
+     * Sets the texture object's magnification and minification filter.
+     * @param mag - Value for the TEXTURE_MAG_FILTER parameter.
+     * @param min - Value for the TEXTURE_MIN_FILTER parameter.
+     * @param bind - Allows to skip binding the texture (e.g., when binding is handled outside).
+     * @param unbind - Allows to skip unbinding the texture (e.g., when binding is handled outside).
+     */
+    @Initializable.assert_initialized()
```

Comment:
Note for the texture refactoring: the texture doesn't really have to be initialized for setting `filter`and `wrap`, right? In C++ it's often done before, so the code doesn't map 1:1.